### PR TITLE
Find go.mod based on Go files

### LINKED ISSUE #5292: Search for `go.mod` whenever if found a Go file
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Go modules lints are run only when `go.mod` is included, not when only some Go files are included.

This is strange when `VALIDATE_ALL_CODEBASE` is set to `true`. A Go module does exist but it's not identified because `go.mod` is not updated.

### Expected Behavior

Go modules lints should run when `VALIDATE_ALL_CODEBASE: false` and some Go files are updated.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by harryzcy**
# Proposed changes

Close #5292

## Readiness checklist

In order to have this pull request merged, complete the following tasks.

### Pull request author tasks

- [ ] I included all the needed documentation for this change.
- [ ] I provided the necessary tests.
- [x] I squashed all the commits into a single commit.
- [x] I followed the [Conventional Commit v1.0.0 spec](https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] I wrote the necessary upgrade instructions in the [upgrade guide](../docs/upgrade-guide.md).
- [x] If this pull request is about and existing issue,
  I added the `Fix #ISSUE_NUMBER` label to the description of the pull request.

### Super-linter maintainer tasks

- [x] Label as `breaking` if this change breaks compatibility with the previous released version.
- [x] Label as either: `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`.

### CODE REVIEW

File: lib/functions/buildFileList.sh [hunk 1]
```diff
@@ -368,6 +368,16 @@ BuildFileArrays() {
       echo "${FILE}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GHERKIN"
     elif [ "${FILE_TYPE}" == "go" ]; then
       echo "${FILE}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO"
+
+      # find go.mod
+      dir_name="${FILE_DIR_NAME}"
+      while [ "${dir_name}" != "$(dirname ${GITHUB_WORKSPACE})" ] && [ "${dir_name}" != "/" ]; do
```
Comment (ferrarimarco) [inline lib/functions/buildFileList.sh:374]: Can you please double-quote `${GITHUB_WORKSPACE}` ?

File: lib/functions/buildFileList.sh [hunk 2]
```diff
@@ -368,6 +368,16 @@ BuildFileArrays() {
...
+        if [ -f "${dir_name}/go.mod" ]; then
+          echo "${dir_name}/go.mod" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO_MODULES"
```
Comment (ferrarimarco) [inline lib/functions/buildFileList.sh:376]: What if we already added this `go.mod` file?

File: lib/functions/buildFileList.sh [hunk 3]
```diff
@@ -368,6 +368,16 @@ BuildFileArrays() {
       echo "${FILE}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GHERKIN"
     elif [ "${FILE_TYPE}" == "go" ]; then
       echo "${FILE}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO"
+
+      # find go.mod
```
Comment (ferrarimarco) [inline lib/functions/buildFileList.sh:372]: Can you add some `debug` statements so we can follow what's happening?

File: lib/functions/buildFileList.sh [hunk 4]
```diff
@@ -368,6 +368,16 @@ BuildFileArrays() {
...
+      dir_name="${FILE_DIR_NAME}"
```
Comment (ferrarimarco) [inline lib/functions/buildFileList.sh:373]: Can you add `local dir_name` before this assignment?

**Comment by ferrarimarco** [comment]
Thanks for this PR! A few small things to check.

File: lib/functions/buildFileList.sh (hunk 2 above)
Comment (harryzcy) [reply lib/functions/buildFileList.sh]: I implemented a solution in 4992fa4, please review it. Thanks

**Comment by ferrarimarco** [comment]
Thanks! A small thing to fix:

```sh
            if [ ! grep -Fxq \\"${go_mod_filename}\\" \\"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO_MODULES\\" ]; then
            ^-- SC1009 (info): The mentioned syntax error was in this if expression.
               ^-- SC1073 (error): Couldn't parse this test expression. Fix to allow more checks.
                                                  ^-- SC1072 (error): Expected test to end here (don't wrap commands in []/[[]]). Fix any mentioned problems and try again.
```

**Comment by harryzcy** [comment]
> Thanks! A small thing to fix:
> 
> ```shell
>             if [ ! grep -Fxq \\"${go_mod_filename}\\" \\"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO_MODULES\\" ]; then
>             ^-- SC1009 (info): The mentioned syntax error was in this if expression.
>                ^-- SC1073 (error): Couldn't parse this test expression. Fix to allow more checks.
>                                                   ^-- SC1072 (error): Expected test to end here (don't wrap commands in []/[[]]). Fix any mentioned problems and try again.
> ```

I have fixed the test

**Comment by ferrarimarco** [comment]
There's an issue, I think:

```
  2024-03-13 21:41:06 [FATAL]   Error when setting INDEX for GO_MODULES: 0
```

https://github.com/super-linter/super-linter/actions/runs/8271680556/job/22632678543?pr=5345

Can you kindly take a look? Thanks!

**Comment by ferrarimarco** [comment]
> There's an issue, I think:
> 
> ```
>   2024-03-13 21:41:06 [FATAL]   Error when setting INDEX for GO_MODULES: 0
> ```
> 
> https://github.com/super-linter/super-linter/actions/runs/8271680556/job/22632678543?pr=5345
> 
> Can you kindly take a look? Thanks!

More logs after rebasing:

```
  2024-04-22 09:49:46 [DEBUG]   PARALLEL_COMMAND_OUTPUT for GO_MODULES (exit code: 255): cd /tmp/lint/test/linters/go_modules/go_modules_good || exit 255; golangci-lint run --allow-parallel-runners -c /tmp/lint/.github/linters/.golangci.yml && echo \"Linted: /tmp/lint/test/linters/go_modules/go_modules_good\"
  parallel: Error: Cannot change into non-executable dir /tmp/lint/test/linters/go_modules/go_modules_good/go.mod: File exists
```

This is due to how we configure the command to run for `GO_MODULES`: https://github.com/super-linter/super-linter/blob/56e675bd333f8af069b2d7ba67ade1f39fcd01c4/lib/functions/worker.sh#L123-L139

So the command becomes:

```sh
parallel ... --workdir {} ...
```

(`{}` means: the input item as it is, which is `/tmp/lint/test/linters/go_modules/go_modules_good/go.mod` because you set it to `echo "${go_mod_filename}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO_MODULES"`).

I'll send another comment with the necessary changes.

Thanks for all your work on this!

**Review by ferrarimarco — CHANGES_REQUESTED** [review]

**Comment by harryzcy** [comment]
@ferrarimarco Thanks for pointing out the issue! It's now updated

**Comment by ferrarimarco** [comment]
Thanks!

We are on the right track!

I think we might still need some refactoring because in the current implementation users will not be able to disable lining of Go modules by using the `VALIDATE_GO_MODULES` variable. Also, we might report false positives because Go files might be linted one by one (`VALIDATE_GO` default mode), and also as a module.

Let me think a bit about this.

File: lib/functions/buildFileList.sh [hunk 5]
```diff
@@ -370,22 +355,43 @@ BuildFileArrays() {
...
+        dir_name="${FILE_DIR_NAME}"
+        local potential_go_mod_file_path="${dir_name}/go.mod"
+        while [ "${dir_name}" != "$(dirname "${GITHUB_WORKSPACE}")" ] && [ "${dir_name}" != "/" ]; do
+          if [ -f "${potential_go_mod_file_path}" ]; then
+            GO_MOD_FILE_PATH="${potential_go_mod_file_path}"
+            break
           fi
-          break
+          dir_name=$(dirname "${dir_name}")
+        done
```
Comment (harryzcy) [inline lib/functions/buildFileList.sh:377]: When walking up the directory, `potential_go_mod_file_path` here would not be changed? It should be updated within the loop for the parent `dir_name`

File: lib/functions/buildFileList.sh [hunk 6]
```diff
@@ -370,22 +355,43 @@ BuildFileArrays() {
...
+        else
+          debug "Skip adding ${GO_MOD_FILE_PATH} directory (${FILE_DIR_NAME}) to the list of Go modules to lint because it's already in that list"
         fi
-        dir_name=$(dirname "${dir_name}")
-      done
+      else
+        debug "${FILE} is not considered to be part of a Go module"
+      fi
+
+      echo "${FILE}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO"
```
Comment (harryzcy) [inline lib/functions/buildFileList.sh:394]: Should line 394 be placed at line 391 instead? i.e. only lint individual Go files when it's not part of a Go module. I'm not sure here

**Comment by harryzcy** [comment]
@ferrarimarco Thank you so much! There're two parts that I have doubts. If you agree, I'll make the edits and squash two commits.

File: lib/functions/buildFileList.sh (hunk 6 above)
Comment (ferrarimarco) [reply lib/functions/buildFileList.sh:394]: In this function, we just list and categorize files. Then, it's up to the user to eventually disable linting for individual Go files if they want to avoid false positives due to scanning individual files when there's a Go module. Before this change, we did checks in this regard, but it looked like a workaround and we had to account for a couple of corner cases about the test environment that complicated the logic when it wasn't necessary.

One check that we need anyway is that this should add `${FILE}` to `"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO"` only if it's a `.go` file, because `${FILE}` can be `go.mod`, and we don't want to add it to `file-array-GO`.

File: lib/functions/buildFileList.sh (hunk 5 above)
Comment (ferrarimarco) [reply lib/functions/buildFileList.sh:377]: This is correct. We should update `potential_go_mod_file_path` after updating `dir_name`. Good catch

**Comment by ferrarimarco** [comment]
Thanks for the review @harryzcy ! Your proposal SGTM :)

File: lib/functions/buildFileList.sh (hunk 6 above)
Comment (harryzcy) [reply lib/functions/buildFileList.sh]: > One check that we need anyway is that this should add `${FILE}` to `"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO"` only if it's a `.go` file, because `${FILE}` can be `go.mod`, and we don't want to add it to `file-array-GO`.

Updated in https://github.com/super-linter/super-linter/pull/5345/commits/cb8ebd42c9805a9f8b97bfacad0238be87e2d014

File: lib/functions/buildFileList.sh (hunk 5 above)
Comment (harryzcy) [reply lib/functions/buildFileList.sh]: Fixed in https://github.com/super-linter/super-linter/pull/5345/commits/cb8ebd42c9805a9f8b97bfacad0238be87e2d014

**Comment by harryzcy** [comment]
I just squashed two commits to become https://github.com/super-linter/super-linter/pull/5345/commits/4637a4ec70525882b1fdee2ccb977e2569d66180

File: lib/functions/buildFileList.sh [hunk 7]
```diff
@@ -370,8 +355,45 @@ BuildFileArrays() {
...
+        else
+          debug "Skip adding ${GO_MOD_FILE_PATH} directory (${FILE_DIR_NAME}) to the list of Go modules to lint because it's already in that list"
+        fi
+      else
+        debug "${FILE} is not considered to be part of a Go module"
+      fi
+
+      if [[ "${BASE_FILE}" != "go.mod" ]]; then
+        echo "${FILE}" >>"${FILE_ARRAYS_DIRECTORY_PATH}/file-array-GO"
+      fi
```
Comment (ferrarimarco) [inline lib/functions/buildFileList.sh:396]: Minor refactoring question. Do we want to move this echo in the else clause (L366) of the if if statement (L363)? That would perhaps make things clearer because we already have the relevant debug statements. WDYT?

Comment (ferrarimarco) [reply lib/functions/buildFileList.sh:396]: @harryzcy

Comment (ferrarimarco) [reply lib/functions/buildFileList.sh]: done

Comment (harryzcy) [reply lib/functions/buildFileList.sh]: thanks
